### ISSUE TITLE
Dealing with boolean attribute selectors

### ISSUE THREAD

**Issue by wonnage** [airbnb/enzyme#197]
Outside of enzyme-land, you'd look for checked checkboxes with `input[checked]`, due to how HTML/CSS handles boolean values. However, in enzyme, this would be wrong, because it would match any checkbox that was passed a `checked` prop (i.e, it'd match every controlled checkbox). You'd want to use `[checked=true|false]` instead.

I wonder if it'd be worth making a change that is aware of prop name selectors potentially being booleans; i.e, `input[checked]` should not match if `checked=false` in props.

**Comment by lelandrichardson** [comment]
cc @ljharb I believe we spent some time discussing this?

**Comment by ljharb** [comment]
tl;dr of our discussion and decision:
A selector string in enzyme should have one of three semantics to make sense:
 - [qSA](http://mdn.io/queryselectorall) semantics: in which `[checked]` matches every string except the empty string (and strings are the only kind of value), and `[foo=3]` thus matches any value that toStrings to `"3"`.
 - JS `===` semantics: in which `[checked]` would match any truthy JS value (meaning, `null`, `undefined`, `false`, `0`, `-0`, `NaN`, and the empty string would be ignored), and `[foo=3]` would match only the number `3`.
 - React semantics: in which `[checked]` would match any value that was not `undefined` - the same logic that React uses to trigger `defaultProps` - and `[foo=3]` would match only the number `3`.

We decided to go with the React semantics for `enzyme`.

In order to achieve your feature request, we'd have to find a confusing middle ground between React, JS, and qSA semantics, where `false` was special and even if you had a `PropTypes.any` prop that passed `false`, it wouldn't match. I don't think that would be an improvement, holistically.

Thanks for the suggestion, though! It's important to continually re-evaluate these kinds of decisions just in case they no longer are valid.
